components/script/dom/readablestream.rs

```diff
@@ -108,426 +255,705 @@ impl ReadableStream {
+            _ => unreachable!(
+                "Enqueueing chunk to a stream from Rust on other than default controller"
+            ),
+        }
+    }
 
-    #[allow(unsafe_code)]
-    pub fn enqueue_native(&self, bytes: Vec<u8>, can_gc: CanGc) {
-        let global = self.global();
-        let _ar = enter_realm(&*global);
-        let cx = GlobalScope::get_cx();
-
-        let handle = unsafe { self.js_stream.handle() };
-
-        self.external_underlying_source
-            .as_ref()
-            .expect("No external source to enqueue bytes.")
-            .enqueue_chunk(cx, handle, bytes, can_gc);
+    /// <https://streams.spec.whatwg.org/#readable-stream-error>
+    pub fn error(&self, e: SafeHandleValue) {
+        // step 1
+        assert!(self.is_readable());
+        // step 2
+        self.state.set(ReadableStreamState::Errored);
+        // step 3
+        self.stored_error.set(e.get());
+
+        // step 4
+        match self.reader {
+            ReaderType::Default(ref reader) => {
+                let Some(reader) = reader.get() else {
+                    // step 5
+                    return;
+                };
+
+                // steps 6, 7, 8
+                reader.error(e);
+            },
+            _ => todo!(),
+        }
     }
 
+    /// <https://streams.spec.whatwg.org/#readablestream-storederror>
     #[allow(unsafe_code)]
-    pub fn error_native(&self, error: Error) {
-        let global = self.global();
-        let _ar = enter_realm(&*global);
-        let cx = GlobalScope::get_cx();
-
+    pub fn get_stored_error(&self, handle_mut: SafeMutableHandleValue) {
         unsafe {
-            rooted!(in(*cx) let mut js_error = UndefinedValue());
-            error.to_jsval(*cx, &global, js_error.handle_mut());
-            ReadableStreamError(
-                *cx,
-                self.js_stream.handle(),
-                js_error.handle().into_handle(),
-            );
+            let cx = GlobalScope::get_cx();
+            self.stored_error.to_jsval(*cx, handle_mut);
         }
     }
 
+    /// <https://streams.spec.whatwg.org/#readable-stream-error>
+    /// Note: in other use cases this call happens via the controller.
     #[allow(unsafe_code)]
-    pub fn close_native(&self) {
-        let global = self.global();
-        let _ar = enter_realm(&*global);
+    pub fn error_native(&self, error: Error) {
         let cx = GlobalScope::get_cx();
+        rooted!(in(*cx) let mut rval = UndefinedValue());
+        unsafe {
+            error
+                .clone()
+                .to_jsval(*cx, &self.global(), rval.handle_mut())
+        };
+        self.error(rval.handle());
+    }
 
-        let handle = unsafe { self.js_stream.handle() };
-
-        self.external_underlying_source
-            .as_ref()
-            .expect("No external source to close.")
-            .close(cx, handle);
+    /// Call into the controller's `Close` method.
+    pub fn close_native(&self) {
```

Comment:
We should rename this to `controller_close_native`, and add a link to `https://streams.spec.whatwg.org/#readable-stream-default-controller-close`.